[Feature] Date input field

esizer opened the pull request:
🤖 Resolves #5645 

## 👋 Introduction

Adds a new `<DateInput />` to the `@gc-digital-talent/forms` package.

## 🕵️ Details

This functions similar to our other inputs except for a few main differences.

### Props

In addition to our typical props for inputs, the `<DateInput />` also accepts the  `show` prop.

```tsx
type DateSegment = "YEAR" | "MONTH" | "DAY"

export interface DateInputProps extends React.HTMLProps<HTMLFieldSetElement> {
  ...
  show?: Array<DateSegment>; // Default: [DATE_SEGMENT.Year, DATE_SEGMENT.Month, DATE_SEGMENT.Day]
}
```

#### `show`

This is an array of the different segments you would like to display. Since we need to send a valid date there is some magic that happens behind the scenes when a segment is not displayed to the user:

 - Year: This is set to the current year
 - Month: Set to January
 - Day: Set to first day of the month 

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Run storybook `npm run storybook`
2. Confirm each story works in regards to the props:
    - Submit correct data
    - Date range validation functions as expected
    - Show/hiding segments works

## 📸 Screenshot

### Desktop
![Screenshot 2023-03-16 131103](https://user-images.githubusercontent.com/4127998/225698697-f4c0b154-21ed-4d99-9857-cfdff452ab3a.png)

### Mobile
![Screenshot 2023-03-16 131113](https://user-images.githubusercontent.com/4127998/225698736-465415b9-f109-47e0-b812-ec54eaf1eb58.png)

petertgiles commented on packages/forms/src/components/DateInput/DateInput.tsx:
I find that the full width is strange.  I don't think these should stretch.
![image](https://user-images.githubusercontent.com/8978655/225389556-0d1e71bc-d3db-4056-bc7e-90eac8cf7926.png)

esizer replied:
@substrae Any thoughts on this? I know you did some styling, just making sure I did not mess something up in a following commit 😢

petertgiles commented on packages/forms/src/components/DateInput/DateInput.tsx:
The mobile view is a bit broken.
![image](https://user-images.githubusercontent.com/8978655/225389767-a6b3c42c-45cf-4566-8fcb-97ee7a3bab77.png)

esizer replied:
Fixed in [fix mobile view](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/09c75af5c411d33c079e3d17c3048a0c7bf1f94e)

petertgiles commented on packages/forms/src/components/DateInput/DateInput.tsx:
I'd like to treat this as a single control so I would like to hide all the decorations on the three boxes.  Also, it's weird to have "optional" and "required" conflicting.
![image](https://user-images.githubusercontent.com/8978655/225390192-3e3bdd65-edd5-4ee0-8b99-0afedf4f82bb.png)

petertgiles replied:
They also cause some alignment problems in French.
![image](https://user-images.githubusercontent.com/8978655/225392557-63e3b85c-d2cb-49a5-9863-bea6b1389966.png)

esizer replied:
Updated in [fix validation, remove individual input rules](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/d509ed7c4f4bb6beec54c765f0d0505f73414745)

petertgiles commented on packages/forms/src/components/DateInput/DateInput.tsx:
Again about treating it as a single control, can we have a single value come back from react-hook-form?  Is this possible with react-hook-form?
![image](https://user-images.githubusercontent.com/8978655/225390795-5fd3d7ca-b68a-4d15-bdfa-2e8912a70502.png)

esizer replied:
I cannot find out how to do this easily from within the input. It looks like [the recommended way](https://github.com/react-hook-form/react-hook-form/discussions/9220#discussioncomment-3891870) is to just remove them in the `onSubmit` -_-

vd1992 replied:
having to edit every single `onSubmit` instance for it definitely won't be fun 😬

esizer replied:
Okay, I think I can sort this out using the [`Controller`](https://react-hook-form.com/api/usecontroller/controller/) component's `render` prop. Going to be a heavy refactor but should be worth it :P 

I think we can render 3 separate inputs in the controller and have it compose the value in the `onChange` event. 🤔 That way we don't need to register the inputs with `react-hook-form`.

vd1992 replied:
oh, seems to emit a single YYYY-MM-DD value now?
that's fantastic and should make implementing so much less daunting

esizer replied:
Yeah, glad we were able to figure that out 😄

petertgiles replied:
This looks so good!

petertgiles commented on packages/forms/src/components/DateInput/DateInput.tsx:
If you hide year, the alignment is strange.
![image](https://user-images.githubusercontent.com/8978655/225391258-c5a25f65-63f7-4e06-b5d3-d18356ab3aca.png)

esizer replied:
Fixed in [fix grid when year is hidden](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/998b9666c85a79dbee54114e4e6211772690b80d)

petertgiles commented on packages/forms/src/components/DateInput/DateInput.stories.tsx:
```diff
@@ -0,0 +1,61 @@
+  idPrefix: "date",
+  legend: "Date",
+  name: "date",
+  dateRange: {
```
It's not possible to use the standard "rules" prop for this?
![image](https://user-images.githubusercontent.com/8978655/225395640-5b40e919-a624-4de6-94ed-b607becb8837.png)

esizer replied:
Hmm, that is interesting. I was under the impression `min` only worked with numbers and not dates 🤔 I'll take a look at this.

esizer replied:
Omg, it works! 🤣 I had no idea!

Added in the refactor: [refactor to use single value Controller](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/2fa762ee94246c30b68f0b9c74ffa1586f627567)

esizer replied:
Okay, looks like it doesnt work 😢 What if we still use the `rules` but extract out `min` and `max` and use it for a custom validate similar to what we were doing before the the `dateRange`? 

That way, we can [use the validate object](https://github.com/react-hook-form/react-hook-form/discussions/8631#discussioncomment-3081122) and still keep things familiar by using `rules.min` and `rules.max`?

esizer replied:
Yeah 👎 Updated in [update to use custom validation](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/ca51c01cb7da5c3e50ba0320448ab3a2f680a2cb)

petertgiles reviewed (commented):
Looking very cool!  I did a quick pass with some first impressions.

petertgiles commented on packages/forms/src/components/DateInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,133 @@
+  };
+
+  const handleYearChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    handleChange(e.target.value, DATE_SEGMENT.Year);
```
Could you pad this one to four digits, too?
![image](https://user-images.githubusercontent.com/8978655/225641539-24236f04-f284-4264-8fcc-43e043af031a.png)

esizer replied:
Padded year in [pad year value to 4 digits](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/95373f4a751381fe160bb88f8ba83d8de76d5415)

petertgiles reviewed (changes requested):
This looks really great!  Two more small things I thought of...

esizer commented on packages/forms/src/components/DateInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,133 @@
+          <label data-h2-display="base(block)" htmlFor={ID.YEAR}>
+            {intl.formatMessage(dateMessages.year)}
+          </label>
+          <input
```
Oops, that is definitely needed. I forgot to carry it over in the refactor 🤦‍♀️

esizer replied:
Added in [add placeholders to inputs](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/37f6f8f22094babca113e10111ced9465a4cde11) and translated (with help from @mnigh ) in [add placeholder translations](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/7956dc5095b0052df16caf20703a86942f563541)

petertgiles commented:
I was thinking about some more "advanced features" that I'd like to see in a story to ensure they work.  A richer story might be good for Fable, too.

- default value coming from an asynchronous source after render, like query would
- validation that depends on another value in the form, like "end date must be after start date"
- watch mode on date value to change another field in the form, like "if after a certain date, hide another form field"

esizer commented:
> I was thinking about some more "advanced features" that I'd like to see in a story to ensure they work. A richer story might be good for Fable, too.

This is an excellent idea 👍 

> * default value coming from an asynchronous source after render, like query would

May need some help on this one, not entirely sure how to go about this in storybook 🤔

petertgiles commented:
> I'm sure we have a story that does this. Let me look...

Yeah, a lot of stories do some asynchronous stuff with promises, like packages/forms/src/components/Combobox/Combobox.stories.tsx.  Something like that might work?

esizer commented:
> Yeah, a lot of stories do some asynchronous stuff with promises, like packages/forms/src/components/Combobox/Combobox.stories.tsx. Something like that might work?

omg, you mean the story I wrote? 🤣 Thanks! I guess I'll copy my own homework 🤫

esizer commented:
> * default value coming from an asynchronous source after render, like query would
> * validation that depends on another value in the form, like "end date must be after start date"
> * watch mode on date value to change another field in the form, like "if after a certain date, hide another form field"

- Added stories for dependent fields in [add input dependent stories](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/59746930d932f10e5346c19a9d6111f7daad259b)
- Added an async default value using a mock database value in [add async story](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/940211a2ce9bb21a39fc8d81866d4aa3f6e850ea)

petertgiles commented on packages/forms/src/components/DateInput/DateInput.stories.tsx:
```diff
@@ -0,0 +1,231 @@
+  );
+};
+
+export const HideInputWhenInvalid = RenderDependantTemplate.bind({});
```
The conditional visiblity seems a little glitchy on this one.

https://user-images.githubusercontent.com/8978655/225688735-e8bb2312-4202-496d-bcd8-d9b827ab2ac6.mp4

esizer replied:
Oops, I was using the wrong condition 🤦‍♀️ Fixed in [update to on submit, fix conditional render story](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/6638dd6b46366b817ea3d488aaab3801773e63bb)

petertgiles commented on packages/forms/src/components/DateInput/DateInput.stories.tsx:
```diff
@@ -0,0 +1,231 @@
+  );
+};
+
+export const Default = Template.bind({});
```
Remember, "no yelling at the user until they've tried to do something wrong". :smile:  I'm not sure how you can "onBlur" when it's three inputs?  Maybe start it off at 0001-01-01 so they start off in a good state?

https://user-images.githubusercontent.com/8978655/225689299-29657848-5dfc-413e-ac61-76587234ccce.mp4

esizer replied:
Looks like the `BasicForm` uses `onChange` by default 🤔 Updated this story to use `onSubmit` in [update to on submit, fix conditional render story](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/6638dd6b46366b817ea3d488aaab3801773e63bb)

petertgiles reviewed (changes requested):
The new stories are so cool!

petertgiles commented on packages/forms/src/components/BasicForm.tsx:
```diff
@@ -55,6 +55,12 @@ function BasicForm<TFieldValues extends FieldValues>({
     ...options,
     defaultValues: options?.defaultValues,
   });
+  console.log({
```
Leftover debugging?

esizer replied:
Oops, yup 👅 

[remove debug statement](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/243eecf527ba43810afc4ad4b01316534b772bb9)

petertgiles reviewed (changes requested):
Looks great!  A leftover console.log but everything else looks good to me.

vd1992 commented:
implemented this change for the pool editing page

![image](https://user-images.githubusercontent.com/40485260/225769977-23b583ec-2142-46ad-9005-0f4c8453ee12.png)

and played around with the page here

https://recordit.co/FKAhNVkImB

seems to work... fantastic? amazing, as usual eric
don't know if the appearance quirks (placeholder text cut off for months) is a concern, might just be resolved by slapping some formatting on in the page file, wrapping the input

vd1992 commented:
oh yeah, and might as well update the index file in packages to export this
then this component is truly completely ready to be swapped in across the site

![image](https://user-images.githubusercontent.com/40485260/225771766-866c7cc7-1b91-4747-9c10-693804d5d856.png)

esizer commented:
> don't know if the appearance quirks (placeholder text cut off for months) is a concern, might just be resolved by slapping some formatting on in the page file, wrapping the input

Thanks @vd1992 , I took a look at this example and it seems to do more with the page styling itself. This is a good thing to be aware of if we implement the component across the application. Looks like this page layout was specific to the `input[type=date]` which is far less wide.

> oh yeah, and might as well update the index file in packages to export this
> then this component is truly completely ready to be swapped in across the site

That is a great idea 😅 Added in [export date input from forms package](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5928/commits/cbd1b206959dceaf3f0ddaad8ea4e52a285837b7)
